### PR TITLE
quic: add qlog support

### PR DESCRIPTION
**Pull request by addaleax**
Fixes: #185

This is as WIP as #177 I assume, as this also requires updating ngtcp2 (and the first commits here are taken from there).

##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

### CODE REVIEW

**Comment by jasnell** [comment]
nice! #177 should be ready to land by Wednesday.

File: doc/api/quic.md [hunk 1]
```diff
@@ -71,6 +71,8 @@ added: REPLACEME
   * `maxConnectionsPerHost` {number} The maximum number of inbound connections
     per remote host. Default: `100`.
   * `port` {number} The local port to bind to.
+  * `qlog` {boolean} Whether to emit ['qlog'][] events for incoming sessions.
+    (For outgoing client sessions, set `client.qlog`.) Default: `false`.
```
Comment (jasnell) [inline doc/api/quic.md]: Does this need to be here or can we base it solely on the presence of the `qlog` event handler the way we do with things like the keylog support?

**Comment by addaleax** [comment]
> nice! #177 should be ready to land by Wednesday (at the latest)

Yeah, there is one more issue here that I think I’ve narrowed down quite a bit … namely, exceptions from nested `MakeCallback`s don’t seem to trigger the fatal exception handler, so exceptions from a `'qlog'` handler currently go unnoticed and the process just continues (and crashes due to some async hooks expectations around uncaught exceptions).

It’s probably not directly related to this PR, but it should be fixed before merging this; I’ve spent some time digging into it yesterday, but it’s not really a trivial issue…

File: doc/api/quic.md (hunk 1 above)
Comment (addaleax) [reply doc/api/quic.md]: @jasnell My understanding is that we need to pass the native qlog callback to nghttp2 when setting up the connection, i.e. when the native handle object is created, and we can’t always know if there’s `'qlog'` event handlers at that point?

Comment (jasnell) [reply doc/api/quic.md]: Unless there is a significant performance hit associated with adding the callback I would always add it and include the logic in the qlog handler to check to see if a listener has been added.

Comment (addaleax) [reply doc/api/quic.md]: @jasnell I’ll try to measure the perf impact, but my personal expectation would be that it’s not trivial.

The bigger issue, though, is that it’s not clear at what point we’d know whether there’s going to be a JS event listener or not, and that we’d have to buffer data up until then – which could be indefinitely, I guess?

**Comment by jasnell** [comment]
Landed #177. This will need a rebase :-)

File: doc/api/quic.md (hunk 1 above)
Comment (jasnell) [reply doc/api/quic.md]: Let's land this as is then but once it lands I'll play around with it a bit more to see if we can reasonably do without the config option.

**Comment by addaleax** [comment]
Yup, rebased :+1:

We can merge this if we want, https://github.com/nodejs/quic/pull/200#issuecomment-558637335 still needs to be addressed but it’s probably going to be mostly unrelated to the contents of this PR anyway

**Comment by jasnell** [comment]
Landed

**Comment by rmarx** [comment]
Oh this is fantastic news! Thank you for the great work!

**Pull request closed without merging**